proxy/http2/Http2Stream.cc

```diff
@@ -273,103 +273,111 @@ Http2Stream::send_request(Http2ConnectionState &cstate)
+      retval = false;
+      break;
 
-  case Http2StreamState::HTTP2_STREAM_STATE_RESERVED_REMOTE:
-    // Currently ATS supports only HTTP/2 server features
-    return false;
+    case Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_LOCAL:
+      if (recv_end_stream) {
+        _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
+      } else {
+        // Error, set state closed
+        _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
+        retval = false;
+      }
+      break;
 
-  case Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_LOCAL:
-    if (type == HTTP2_FRAME_TYPE_RST_STREAM || recv_end_stream) {
-      _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
-    } else {
-      // Error, set state closed
-      _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
-      return false;
-    }
-    break;
+    case Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_REMOTE:
+      if (send_end_stream) {
+        _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
+      } else if (type == HTTP2_FRAME_TYPE_HEADERS) { // w/o END_STREAM flag
+        // No state change here. Expect a following DATA frame with END_STREAM flag.
+      } else if (type == HTTP2_FRAME_TYPE_CONTINUATION) { // w/o END_STREAM flag
+        // No state change here. Expect a following DATA frame with END_STREAM flag.
+      } else {
+        // Error, set state closed
+        _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
+        retval = false;
+      }
+      break;
 
-  case Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_REMOTE:
-    if (type == HTTP2_FRAME_TYPE_RST_STREAM || send_end_stream) {
-      _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
-    } else if (type == HTTP2_FRAME_TYPE_HEADERS) { // w/o END_STREAM flag
-      // No state change here. Expect a following DATA frame with END_STREAM flag.
-      return true;
-    } else if (type == HTTP2_FRAME_TYPE_CONTINUATION) { // w/o END_STREAM flag
-      // No state change here. Expect a following DATA frame with END_STREAM flag.
-      return true;
-    } else {
-      // Error, set state closed
-      _state = Http2StreamState::HTTP2_STREAM_STATE_CLOSED;
-      return false;
+    case Http2StreamState::HTTP2_STREAM_STATE_CLOSED:
+      // No state changing
+      break;
+    default:
+      retval = false;
+      break;
     }
-    break;
-
-  case Http2StreamState::HTTP2_STREAM_STATE_CLOSED:
-    // No state changing
-    return true;
+  }
 
-  default:
-    return false;
+  // Did we just transition to closed or remote closed? Decrement stream count
+  if (_state == Http2StreamState::HTTP2_STREAM_STATE_CLOSED && original_state != Http2StreamState::HTTP2_STREAM_STATE_CLOSED) {
+    Http2ClientSession *h2_proxy_ssn = static_cast<Http2ClientSession *>(this->_proxy_ssn);
+    SCOPED_MUTEX_LOCK(lock, h2_proxy_ssn->connection_state.mutex, this_ethread());
+    h2_proxy_ssn->connection_state.decrement_stream_count(this->_id);
```

Comment:
I can understand we can check state changes here, but I'm not a big fan of giving another task to the `Http2Stream::change_state()`.